### ISSUE TITLE
Add Support for "Ancient" .smc Payload Files Without Adler-32

### ISSUE THREAD

**Issue by microwave89-hv** [sadponyguerillaboy/SMC-Toolkit#1]
Currently, the SMC-Toolkit is stated to only support .smc files for EC's in 2013 years HW
It would be nice to add support for very old ("ancient") .smc payloads.
The key difference of ancient .smc payloads, such as of a MBP10,1 (MacBook Pro Retina 15" Mid-2012), is the missing adler-32 sums.
This means, that the app region's checksum is not checked by the SMC updater on flashing a new app.
The only validation mechanism is the checksum bytes at the end of each line.

If the SMC-Toolkit is still used to reconstruct the firmware.bin image from an ancient .smc update then all region types are stated to have checksum errors.
Extraction worked as expected though since the file format is the same as in the old .smc updates.
Creation of a new ancient payload hasn't been attempted but is believed to fail.
With support of ancient .smc files the toolkit would no longer calculate the adler-32 sums when it knows that the current payload is an ancient one. It would stick with calculating the headers and checksum bytes at the line ends.

The information whether the current payload is ancient could be gained either from a command line switch (--ancient) or from previously trying to calculate the adler-32 directly from the .smc file prior to extracting, and then maybe setting an ancient flag if the adler-32 is already failing, possibly with consent of the user ("File recognized as ancient! Are you sure that this is a pre-2013 file?")

As an alternative an attempt could be made to leverage the payload version numbers, and setting an ancient flag if the current number is below a certain threshold.

Please find the ancient .smc update fresh from my own machine.
High Sierra MBP10,1 (https://github.com/microwave89-hv/SMC-Toolkit/blob/main/SMCPayloads/10.13.6/MBP10%2C1/2012MBPR15.smc)

Kind Regards

**Comment by sadponyguerillaboy** [comment]
Is definitely something to consider. My only question would be, is the missing regions in the smc file you linked (`D:00000800 - D:00006000`) filled with `xFF` like the newer variations? If you look at Alex Ionescu's Ninjas and Harry Potter presentation, he notes that regions `D:00000800 - D:00005800` are unknown or unused (which makes sense in this case), but that region `D:00006000` (also missing in your payload) is reserved for the EPM UV (Checksums). Not sure if that missing region is also filled with `xFF` or just not provided? Haven't spent enough time with these older versions to know for certain. Do they require a checksum at all to flash? Does there exist an overall checksum of the entire payload to verify that the assembled version is indeed correct (a benefit of the adler32 sums)? Alex Ionescu's 2014 presentation Smc the Place to Be discusses patching the MACA function to allow the MACM to read out the entire smc in 32 byte chunks. Not sure if this is also applicable to these older versions?

**Comment by microwave89-hv** [comment]
Thank you for your response and your interesting findings!
Incidentally I was curious about the entire dump after I watched the talk a few times, and hence I did exactly the patch that you have suggested.
Please find the 256 KiB dmp here: https://github.com/microwave89-hv/SMC-Toolkit/blob/main/SMCPayloads/10.13.6/Mac-C3EC7CD22292981F/MBP101smc.rom
(Note that in the meantime I also found my own board ID, which is why my directory now is looking like the existing ones.)

As for the 02030f00 in the last data line: I am pretty confident that it only refers to the (App) version.
My "hello world" before flashing the MACA patch was to alter that number and it did show a different version after flashing.
We can try to patch further values and see what we get by these keys:

* RBr          ch8* 		K_DESC_STR 	Source branch
* REV 	{rev 		K_CONST 	Source revision
* RPlt 	ch8* 		K_CONST 	Platform String
* RSvn 	ui32 	4 	K_CONST 	SVN database revision

As far as my experiments went there was never any in-data checksum unlike with the newer files.

**Comment by sadponyguerillaboy** [comment]
email me. you can get it in my profile. thanks

**Comment by sadponyguerillaboy** [comment]
Update: New tools have been released that support the dumping of a patched 2012+ firmware. Further research is required for older Renesas H8S/2117 chips.

**Comment by microwave89-hv** [comment]
Using the tool "reconstruct.py" along with "custompayload.py", a MBPR 2012 .smc payload could now be handled successfully.
The sha256 hashes of both the "reconstruct.py"-firmware.bin and and the "custompayload.py"-custom_payload.smc have been verified successfully.
Despite other things haven't been tested, this particular issue seems to be solved and might be closed in my opinion.

**Comment by sadponyguerillaboy** [comment]
@microwave89-hv agreed about the H8S/2117. I think it would be a reasonable goal to work towards. So will leave this open for now.

**Comment by protocold** [comment]
I am trying to get a smc dump for iMac 2010. Is there a chance that this could be done?

**Comment by microwave89-hv** [comment]
Hi there

Unfortunately, I'm mentally far away from that stuff. How long are you able to wait?
Can you help me speed up the process by answering the following questions?

- What have you tried so far?
- What is preventing you from getting that dump in the first place?
- Or do you have the dump but you cannot decode it?

Right now, I don't even know why I'm on that mailing list, and what problem it was that we tried to solve here (how is dumping related to Adler-32 or such)...
I only know that I haven't touched up on the SMC matter for more than a year now...
Will have to give me some days, or even weeks, depending on what exactly the problem is you have.

Kind regards
microwave89-hv

Sent with [ProtonMail](https://protonmail.com/) secure email.

------- Original Message -------

**Comment by protocold** [comment]
sorry i am a noob... I havent actually tried anything just yet as I follow this issue that you opened over a year ago "Add Support for "Ancient" .smc Payload Files Without Adler-32" [which is probably why you are on this mailing list].

All I am trying to do is to get a dump from a H8S/2117 and then write the dump into another H8S/2117..

Is there anything that I should just go try now?

**Comment by microwave89-hv** [comment]
I'm sorry for my unresponsiveness :/

I will try to come back to you this week, with some suggestions. I haven't actually dumped old SMCs yet, since I have no old Macs lying around. Have you already had a read on that? "http://publications.alex-ionescu.com/NoSuchCon/NoSuchCon%202013%20-%20Ninjas%20and%20Harry%20Potter%20-%20Spellunking%20in%20the%20Apple%20SMC%20Land.pdf"

Tells me that you cannot easily dump the SMC in the first place though...? Maybe you need some SMC update first of all, then you can try to make it into binary, and see if you can create a new update from it.
Or flash the update onto the target SMC, directly.

Really, I need some time, to remember what is possible to be able to help...
I will get back to you soonest.

Kind regards,
microwave89-hv

P.S. As I'm writing, I dimly remember that likely, I was able to read at least large portions of the ROM. I think, what I must have done back in the times was changing the allowed MACA range, then flashing the patched update and thus reading the formerly inaccessible memory regions.
I reckon you would have to try in a similar way...

Sent with [ProtonMail](https://protonmail.com/) secure email.

------- Original Message -------

**Comment by microwave89-hv** [comment]
You just made me wondering what your problem really is, and how it can be sorted.

I just read up on what we were doing last year, to figure the root cause of your problem.
Apparently, Sad Pony had succeeded in adding support for ancient dumps, to be flashed though onto new (LM4F) HW only.
What your problem is, is that you have old HW, as you already have mentioned.

First of all, have you an SMC update firmware file, compatible with your iMac 2010?

Secondly, could you please report your current SMC firmware's version number of the H8/2117 you want to dump? You can find it (at least on High Sierra) under "Apple icon" ==> "About this Mac" ==> "System Report...", "SMC version (system)". It should look something like this "1.53f13".

Thirdly, please provide a little context on your exact issue. Why do you want to clone your iMac 2010 SMC firmware to another H8/2117? Why not just update the target H8/2117 using an Apple-provided SMC update for your target system? How did you come to think you'd have to flash your target system's SMC in the first place? Do you believe you have issues with power management?

Comprehensive feedback very much appreciated :)

Kind regards,
microwave89-hv

Sent with [ProtonMail](https://protonmail.com/) secure email.

------- Original Message -------

**Comment by Electro-it-lab** [comment]
Please make bin file for me I don't know about coding , i need bin file for MacBook pro 9,2  , 2012mbp13.smc , I need this bin file for direct debugging  SMC chips using j-link JTAG

**Comment by MLB-Apple** [comment]
> Por favor, crie um arquivo bin para mim, não sei sobre codificação, preciso de um arquivo bin para MacBook pro 9,2, 2012mbp13.smc, preciso deste arquivo bin para depuração direta de chips SMC usando j-link JTAG

Hi, do you still need the SMC file for your MacBook pro 9.2 A1286 2012?
